Predict when the next reward will be farmed

abhi3700 opened the pull request:
## Background

The [issue](https://github.com/subspace/space-acres/issues/86) titled "Predict when the next reward will be farmed" discusses the potential to provide users (farmers) with predictions on when they can expect to receive rewards based on on-chain data regarding space pledged and the last farming reward payment timestamp. This prediction could then be displayed to the user as having a high/low probability of receiving a reward with descriptions like "today", "in about an hour", or "any time now".

## Description

So, this PR adds the feature to calculate the expected time in getting the next reward (block/vote) as a part of backend.

The calculation is based on the amount of one's space pledged and total space pledged & time elapsed since the last reward timestamp.
> If there is no reward payout yet, then we the current timestamp is multiplied with the inverse of contribution ratio, where `contribution_ratio` = `total_space_pledged`/`space_pledged`.

Details on [Notion](https://www.notion.so/subspacelabs/Farmer-s-Reward-Prediction-Space-Acres-f420a82c67324028a2277affafd88e6e?pvs=4).

<img width="697" alt="image" src="https://github.com/subspace/space-acres/assets/16472948/3fab95c0-6c94-4f86-99ee-4e78f7070c90">

There are 2 main functions:
- [`get_total_space_pledged`](https://github.com/subspace/space-acres/blob/b630ae745b174aae9e3283c31aa7a798cd2adab5/src/backend/node.rs#L346-L371): This function returns the total space pledged as per your latest block available in your synced node.
- [`calculate_expected_reward_duration_from_now`](https://github.com/subspace/space-acres/blob/b630ae745b174aae9e3283c31aa7a798cd2adab5/src/backend/farmer.rs#L579-L594): This function returns the intended expected time to get the next reward.
  Now this function can be integrated with frontend UI.

## Appendix

Here, the challenge is to fetch the current solution range, runtime constants like Slot probability.
Initially, I attempted with `graphql`, but due to the fact that it always lags behind 10 blocks in realtime data fetching, so switched to RPC approach using `subxt` crate. And then I finally switched to runtime API call as the former added technical debt.

> As of https://github.com/subspace/space-acres/pull/161/commits/02c8c7a45eb91e50e3c7bf1665b6ae7f1d0f3791, the 3rd approach is implemented.

The 4th approach is:
- [ ] to get the `total_space_pledged` directly as runtime constants without any calculation.

nazar-pc reviewed (changes requested):
Also this seems to be some WIP version of the PR due to `#[allow(dead_code)]` and no UI changes

abhi3700 commented on Cargo.toml:
```diff
@@ -91,12 +91,15 @@ subspace-rpc-primitives = { git = "https://github.com/subspace/subspace", rev =
 subspace-runtime-primitives = { git = "https://github.com/subspace/subspace", rev = "b96ae888720551b22ece25cf1adba00615721fa9" }
 subspace-service = { git = "https://github.com/subspace/subspace", rev = "b96ae888720551b22ece25cf1adba00615721fa9" }
+# connect to subspace node as client to make runtime API calls
+subxt = { version = "0.34" }
```
Added the changes.

abhi3700 commented:
As of now, removed RPC based code i.e. `subxt` related code snippets. And replaced with runtime API call instead, fetching `current solution range` value. And then have to calculate `total_space_pledged`. Pushed code.

In order to get the `total_space_pledged` value directly, I pushed changes to [branch](https://github.com/subspace/subspace/tree/abhi3700/add-tsp-runtime-api) to declare & implement runtime APIs to `subspace-runtime`.
If the `runtime_api` implementation is correct, then I can access the `total_space_pledged` value directly without any calculations.

nazar-pc commented on src/backend/node.rs:
```diff
@@ -280,6 +282,41 @@ fn get_total_account_balance(
 }
 
+// defined here: https://github.com/subspace/subspace/blob/5d8b65740ff054b01ebcbaf5a905e74274c1a5d0/crates/subspace-core-primitives/src/pieces.rs#L803
+const PIECE_SIZE: usize = 1048672/* the actual piece size from your runtime */;
```
Piece size does not depend on runtime in any way, you can just use it as `Piece::SIZE`

abhi3700 replied:
Fixed in commit@https://github.com/subspace/space-acres/commit/bd5fbd77b7ebd8b8073c93562d2ac8437f11f5ee

nazar-pc commented on Cargo.toml:
```diff
@@ -115,58 +116,11 @@ numa = [
 #
 # This list is ordered alphabetically.
 [profile.dev.package]
-bitvec = { opt-level = 3 }
```
These changes should not be committed

abhi3700 replied:
Fixed in commit@https://github.com/subspace/space-acres/commit/bd5fbd77b7ebd8b8073c93562d2ac8437f11f5ee

nazar-pc commented on Cargo.toml:
```diff
@@ -97,6 +97,7 @@ thread-priority = "0.16.0"
 tokio = { version = "1.36.0", features = ["fs", "time"] }
 tracing = "0.1.40"
 tracing-subscriber = "0.3.18"
+sp-api = "28.0.0"
```
Every single dependency until this point was sorted alphabetically, please preserve that.

abhi3700 replied:
Fixed in commit@https://github.com/subspace/space-acres/commit/bd5fbd77b7ebd8b8073c93562d2ac8437f11f5ee

nazar-pc commented on src/backend/node.rs:
```diff
@@ -280,6 +282,41 @@ fn get_total_account_balance(
+        .runtime_api()
+        .solution_ranges(block_hash)
+        .map(|solution_ranges| solution_ranges.current)
+        .expect("Failed to get current solution range");
```
This is not really a proof. Check other usages of `.expect()`.

abhi3700 replied:
Not sure, if you meant this.
I refactored to:
```diff
-let current_solution_range = client
-        .runtime_api()
-        .solution_ranges(block_hash)
-        .map(|solution_ranges| solution_ranges.current)?;
+ let current_solution_range = client
+        .runtime_api()
+        .solution_ranges(block_hash)
+        .map_err(|_| "Failed to get solution ranges")?
+        .current;

nazar-pc replied:
This is what I meant: https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md#code-style

Error handling is fine, though your example above hides the actual error, so it will be hard to debug when/if it happens.

abhi3700 replied:
After reading it, what I got to know is...we should avoid panicking as much as possible.
Can this code be considered:
```rust
let current_solution_range = match runtime_api
        .solution_ranges(block_hash)
        .map(|solution_range| solution_range.current)
    {
        Ok(res) => res,
        Err(_) => {
            return Err(ApiError::UnknownBlock(
                "Failed to get solution ranges at this block".to_string(),
            ))
        }
    };

let slot_probability = runtime_api
    .chain_constants(block_hash)
    .map(|chain_constants| chain_constants.slot_probability())
    .unwrap_or((1, 6));

let max_pieces_in_sector = runtime_api
    .max_pieces_in_sector(block_hash)
    .unwrap_or(1000_u16);
 ```
 
 assuming to set a default value of `slot_probability`, `max_pieces_in_sector`.

nazar-pc commented on src/backend/node.rs:
```diff
@@ -280,6 +282,41 @@ fn get_total_account_balance(
+        .map(|solution_ranges| solution_ranges.current)
+        .expect("Failed to get current solution range");
+
+    // Calculate the total space pledged
+    let total_space_pledged = u128::from(u64::MAX)
+        .checked_mul(PIECE_SIZE as u128)
+        .expect("Multiplication with piece size works; qed")
+        .checked_mul(u128::from(SLOT_PROBABILITY.0))
+        .expect("Multiplication with slot probability_0 works; qed")
+        .checked_div(u128::from(current_solution_range))
+        .expect("Division by current solution range works; qed")
+        .checked_div(u128::from(SLOT_PROBABILITY.1))
+        .expect("Division by slot probability_1 works; qed");
```
Where did you get this formula though? Runtime logic has a different structure without `u128`, etc.

abhi3700 replied:
Reverted to [original](https://github.com/subspace/subspace/blob/df8d33b65fff6a88d77fa8090533879199bcb422/test/subspace-test-runtime/src/lib.rs#L375-L381).
Actually I had added checked arithmetic on top of this.

abhi3700 replied:
I see.
Let me change the formula then.

abhi3700 replied:
Fixed [here](https://github.com/subspace/space-acres/blob/1a542daef6241672e2007df4a39815d8155b927c/src/backend/node.rs#L289-L318).

Should I use `SLOT_PROBABILITY` constant, fetched via runtime_api [code](https://github.com/subspace/space-acres/blob/1a542daef6241672e2007df4a39815d8155b927c/src/backend/node.rs#L305-L310) or define as constant like [this](https://github.com/subspace/space-acres/blob/1a542daef6241672e2007df4a39815d8155b927c/src/backend/node.rs#L286).

nazar-pc replied:
Query both slot probability and `MAX_PIECES_IN_SECTOR` using runtime API. Slot probability can also be cached since we don't expect it to ever change, max pieces in sector may change though.

abhi3700 replied:
Fixed [here](https://github.com/subspace/space-acres/commit/31c03ebeeb777e923c047f631653fbf505518fd2).

abhi3700 commented:
> ....did you look into a way to query runtime constants are all?

I searched in substrate docs, paritytech docs for `sp-api`, etc. But couldn't find a way to query the runtime constants.
Everywhere, it was mentioned either to use RPC (via `subxt`) to get the runtime constant or add constant as a function to runtime_api.

nazar-pc commented on src/backend/utils.rs:
````diff
@@ -1,5 +1,26 @@
+// https://github.com/subspace/subspace/blob/df8d33b65fff6a88d77fa8090533879199bcb422/crates/subspace-runtime/src/lib.rs#L104
+pub(crate) const MAX_PIECES_IN_SECTOR: u16 = 1000;
+
+/// Computes the following:
+/// ```
+/// MAX * slot_probability / (pieces_in_sector * chunks / s_buckets) / solution_range
+/// ```
+pub(crate) const fn solution_range_to_sectors(solution_range: SolutionRange) -> u64 {
+    let sectors = SolutionRange::MAX
+        // Account for slot probability
+        / SLOT_PROBABILITY.1 * SLOT_PROBABILITY.0
+        // Now take sector size and probability of hitting occupied s-bucket in sector into account
+        / (MAX_PIECES_IN_SECTOR as u64 * Record::NUM_CHUNKS as u64 / Record::NUM_S_BUCKETS as u64);
+
+    // Take solution range into account
+    sectors / solution_range
+}
````
This code should have a TODO pointing to source code and saying that it needs to be de-duplicated.

abhi3700 replied:
Fixed [here](https://github.com/subspace/space-acres/commit/31c03ebeeb777e923c047f631653fbf505518fd2).

abhi3700 commented on src/backend/utils.rs:
```diff
@@ -1,5 +1,26 @@
 pub(super) type HandlerFn<A> = Arc<dyn Fn(&A) + Send + Sync + 'static>;
 pub(super) type Handler<A> = Bag<HandlerFn<A>, A>;
+
+// defined as constant in `subspace-farmer-components`, `subspace-runtime`.
```
Fixed here.

abhi3700 commented on Cargo.toml:
```diff
@@ -168,6 +169,7 @@ x25519-dalek = { opt-level = 3 }
 yamux = { opt-level = 3 }
 zeroize = { opt-level = 3 }
 
+
```
Fixed [here](https://github.com/subspace/space-acres/commit/31c03ebeeb777e923c047f631653fbf505518fd2).

abhi3700 commented on Cargo.toml:
```diff
@@ -77,6 +77,7 @@ semver = "1.0.22"
 serde = { version = "1.0.197", features = ["derive"] }
 serde_json = "1.0.114"
 simple_moving_average = "1.0.2"
+sp-api = "28.0.0"
```
Fixed [here](https://github.com/subspace/space-acres/commit/31c03ebeeb777e923c047f631653fbf505518fd2).

abhi3700 commented:
In terms of frontend, do you think this would look nice here?

The green colored text.

<img width="613" alt="image" src="https://github.com/subspace/space-acres/assets/16472948/5417e797-9829-4c60-8a44-f558518697d6">

nazar-pc commented:
> In terms of frontend, do you think this would look nice here?

Hard to say, but for sure not where the green text for sure. That whole area is for farms and has scrolling. Should be somewhere around tSSC. Something like [this](https://www.jqueryscript.net/images/Dynamic-Circular-Progress-Bar-with-jQuery-CSS3.jpg) would be nice, but it would require rendering manually with `plotters` library, so probably tiny https://docs.gtk.org/gtk4/class.ProgressBar.html to start is sufficient, which has `[pulse-step](https://docs.gtk.org/gtk4/property.ProgressBar.pulse-step.html)` property.

> 💭 Should we attach the webpage url as well, that shows rewards distributed to all the addresses?..may be block no. or so where the rewards were distributed. I guess that would require querying the synced node 🤔 .

Don't think I understood what you wanted to say here.

abhi3700 commented:
Okay, I will look into it.

> Don't think I understood what you wanted to say here.
I was basically suggesting a feature where a farmer could query for all the address in the rewards list. May be putting a link to the rewards page in block explorer.

nazar-pc commented:
> I was basically suggesting a feature where a farmer could query for all an address in the rewards list. May be putting a link to the rewards page in block explorer would work. Not sure. Need to think on this.

Still didn't understand what you meant. Did you try to click on the balance yet, is that what you meant?

abhi3700 commented:
> > I was basically suggesting a feature where a farmer could query for all an address in the rewards list. May be putting a link to the rewards page in block explorer would work. Not sure. Need to think on this.
> 
> Still didn't understand what you meant. Did you try to click on the balance yet, is that what you meant?

Nevermind, I get it.

nazar-pc commented:
I'm closing this PR since we can't merge it as is. We can reopen it later once both frontend and backend parts are implemented, tested and organized cleanly in a state that is ready for review.

nazar-pc commented:
There are 32 commits here, may I ask you to reorganize PR according to https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md?
It should have at most a few cleanly organized commits that can be reviewed individually, no way I'll be going through 32 commits worth of changes for so few lines of code.
